File: darkturo/perl/string_calculator_test.pl
use Test;
use string_calculator;

BEGIN {
    plan( tests => 13 )
}

ok (0, add(""), "Test with empty string");

ok (1, add("1"), "Test with one number");

ok (3, add("1,2"), "Test call with two numbers");

ok (6, add("1,2,3"), "Unknown amount of numbers");

ok (15, add("1,2,3,4,5"), "Unknown amount of numbers second try");

ok (6, add("1\n2,3"), "new lines between numbers");

ok (3, add("//;\n1;2"), "change delimiter to ';'");

ok ("negatives not allowed: -1", add("-1"), 
    "negatives not allowed, only one negative");

ok ("negatives not allowed: -1, -3, -5", add("-1,2,-3,4,-5"), 
    "negatives not allowed, only one negative");

ok (2, add("2,1001"), "Numbers bigger than 1000 should be ignored");

ok (6, add("//[***]\n1***2***3"), "Delimiters can be of any length");

ok (6, add("//[*][%]\n1*2%3"), "Allow multiple delimiters");

ok (15, add("//[***][;;;][##]\n1***2;;;3;;;4##5"), 
    "handle multiple delimiters with length longer than one char")

File: discolo/scala/src/test/scala/StringCalculatorTest.scala


import org.scalatest.FunSuite
import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import StringCalculator._
import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner

@RunWith(classOf[JUnitRunner])
class StringCalculatorTest extends FunSuite {

  test("add empty string") {
    assert(add("") === 0)
  }

  test("add one number") {
    assert(add("3") === 3)
  }
  
  test("add two numbers") {
    assert(add("4,5") === 4+5)
  }
    
  test("add any numbers") {
    assert(add("4,5,6,7,8,9") === 4+5+6+7+8+9)
  }
  
  test("new lines as delimiters") {
    assert(add("5,6\n7") === 5+6+7)
  }
  
  test("new lines as delimiters - invalid case") {
    try {
      add("5,6\n,7")
      fail()
    } catch {
      case e: NumberFormatException => assert(true)
    }
  }
  
  test("set new delimiter") {
    assert(add("//;\n9;99;999") === 9+99+999)
  }
  
  test("negatives not allowed") {
    try {
      add("5,-6\n7")
      fail()
    } catch {
      case e: Error => assert(true)
    }
  }
  
  test("numbers greater than 1000 are ignored") {
    assert(add("5,1001,6") === 5 + 6)
  }
  
  test("number 1000 is not ignored") {
    assert(add("5,1000,6") === 5 + 1000 + 6)
  }
  
  test("multilength delimiter") {
    assert(add("//[aaa]\n4aaa5aaa1001aaa4") === 4 + 5 + 4)
  }
  
  test("multilength delimiter with special characters") {
    assert(add("//[_*_]\n4_*_5_*_1001_*_4") === 4 + 5 + 4)
  }
  
  test("multilength multiple delimiter") {
    assert(add("//[_*_][***]\n4_*_5***1001_*_4") === 4 + 5 + 4)
  }
}